crates/sparrow-runtime/src/prepare.rs

```diff
@@ -140,192 +137,92 @@ impl<'a> fmt::Display for SourceDataWrapper<'a> {
-
-            let prepare_object_store_url = ObjectStoreUrl::from_str(prepared_url.as_str())
-                .change_context_lazy(|| Error::InvalidUrl(prepared_url.as_str().to_string()))?;
-            let metadata_object_store_url =
-                ObjectStoreUrl::from_str(metadata_prepared_url.as_str()).change_context_lazy(
-                    || Error::InvalidUrl(metadata_prepared_url.as_str().to_string()),
-                )?;
-
-            parquet_uploads.push(
-                object_store_registry
-                    .upload(prepare_object_store_url, Path::new(&prepared_metadata.path)),
-            );
-
-            parquet_uploads.push(object_store_registry.upload(
-                metadata_object_store_url,
-                Path::new(&prepared_metadata.metadata_path),
-            ));
-
-            let result_prepared_file: PreparedFile = prepared_metadata
-                .to_owned()
-                .with_path(prepared_url)
-                .with_metadata_path(metadata_prepared_url)
+        // Create the prepared file via PreparedMetadata.
+        // TODO: We could probably do this directly, eliminating the PreparedMetadata struct.
+        let prepared_file: PreparedFile =
+            PreparedMetadata::try_from_data(data_url.to_string(), &data, metadata_url.to_string())
+                .into_report()
+                .change_context(Error::Internal)?
                 .try_into()
                 .change_context(Error::Internal)?;
-            prepared_files.push(result_prepared_file)
-        }
+        prepared_files.push(prepared_file);
+
+        uploads.push(write_parquet(data, data_url, object_store.clone()));
+        uploads.push(write_parquet(metadata, metadata_url, object_store.clone()));
     }
 
-    // TODO: We could (instead) use a loop and select, which would allow us
-    // to fail early if anything failed. But it is more book-keeping.
-    let parquet_uploads = futures::future::join_all(parquet_uploads).in_current_span();
-    for result in parquet_uploads.await {
-        result.change_context(Error::UploadResult)?;
+    // Wait for the uploads.
+    while let Some(upload) = uploads.try_next().await? {
+        tracing::info!("Finished uploading {upload}");
     }
 
     Ok(prepared_files)
 }
 
+async fn write_parquet(
+    batch: RecordBatch,
+    url: ObjectStoreUrl,
+    object_store: Arc<dyn ObjectStore>,
+) -> error_stack::Result<ObjectStoreUrl, Error> {
+    let path = url
+        .path()
+        .change_context_lazy(|| Error::Write(url.url().clone()))?;
+    let (upload_id, writer) = object_store
+        .put_multipart(&path)
+        .await
+        .into_report()
+        .change_context_lazy(|| Error::Write(url.url().clone()))?;
+    tracing::info!("Multipart upload to {url} started with ID {upload_id}");
+
+    let mut writer = parquet::arrow::AsyncArrowWriter::try_new(writer, batch.schema(), 1024, None)
```

Comment:
Is a `1024` an arbitrary buffer size or something we've confirmed is a good buffer size? https://arrow.apache.org/rust/parquet/arrow/async_writer/struct.AsyncArrowWriter.html#method.try_new

Reply:
Arbitrary.

Based on:

> The intermediate buffer will automatically be resized if necessary.

It seems unlikely it will be a problem. But, we'd likely benefit from larger buffer sizes (although that would limit the number of concurrent uploads we could do). 

S3 seems to suggest it needs to be at least 5 MB (although that may be the "resized if necessary". We could start there? Thoughts on a good starting place?

Reply:
Updated to a constant and set to 5 MB. We can adjust in future PRs after measuring.